panic booting an s10 zone with DEBUG bits

```
panic[cpu14]/thread=fffffe2d84560880: assertion failed: p->p_brand_data == NULL, file: ../../common/os/brand.c, line: 1191

fffffe004040fac0 fffffffffbdcde25 ()
fffffe004040faf0 unix:brand_solaris_setbrand+5a ()
fffffe004040fb10 s10_brand:s10_setbrand+18 ()
fffffe004040fb50 unix:brand_setbrand+112 ()
fffffe004040fdd0 genunix:exec_common+c65 ()
fffffe004040fe60 genunix:exec_init+28a ()
fffffe004040fe80 genunix:start_init_common+f5 ()
fffffe004040fee0 genunix:zone_start_init+8c ()
fffffe004040fef0 unix:thread_start+b ()


panic: entering debugger (continue to save dump)

Welcome to kmdb
kmdb: unable to determine terminal type: assuming `vt100'
Loaded modules: [ scsi_vhci stmf_sbd crypto mac cpc uppc neti ptm ufs unix zfs krtld s1394 sata apix uhci hook lofs genunix ip logindmux nsmb usba xhci specfs pcplusmp nfs random mm cpu.generic arp stmf vmm sockfs smbsrv ]
[14]> $C
fffffffffbc9ce60 kmdb_enter+0xb()
fffffffffbc9ce90 debug_enter+0x75(fffffffffb97a788)
fffffffffbc9cf80 panicsys+0x56e(fffffffffbf27cd8, fffffe004040fa18, fffffffffbc9cf90, 1)
fffffe004040fa00 vpanic+0x15c()
fffffe004040fa70 0xfffffffffb8c2361()
fffffe004040fac0 0xfffffffffbdcde25()
fffffe004040faf0 brand_solaris_setbrand+0x5a(fffffe2dbc0d0fc0, ffffffffc00d7940)
fffffe004040fb10 s10_brand`s10_setbrand+0x18(fffffe2dbc0d0fc0)
fffffe004040fb50 brand_setbrand+0x112(fffffe2dbc0d0fc0, 1)
fffffe004040fdd0 exec_common+0xc65(8047ff4, 8047fe8, 0, 2)
fffffe004040fe60 exec_init+0x28a(fffffe2da0fdf2f8, fffffe2d754f4420)
fffffe004040fe80 start_init_common+0xf5()
fffffe004040fee0 zone_start_init+0x8c()
fffffe004040fef0 thread_start+0xb()
[14]> fffffe2dbc0d0fc0::print proc_t p_brand_data
p_brand_data = 0xfffffe2db9f33580
[14]> 0xfffffe2db9f33580::whatis
fffffe2db9f33580 is allocated from kmem_alloc_64:
            ADDR          BUFADDR        TIMESTAMP           THREAD
                            CACHE          LASTLOG         CONTENTS
fffffe2db875b460 fffffe2db9f33580        bcd3ef2b4 fffffe2d84560880
                 fffffe2cfb430648 fffffe2d145866c0 fffffe2d24972448
                 kmem_cache_alloc_debug+0x37c
                 kmem_cache_alloc+0x138
                 kmem_zalloc+0x4a
                 brand_setbrand+0x166
                 exec_common+0xc65
                 exec_init+0x28a
                 start_init_common+0xf5
                 zone_start_init+0x8c
                 thread_start+0xb

[14]> $C
fffffffffbc9ce60 kmdb_enter+0xb()
fffffffffbc9ce90 debug_enter+0x75(fffffffffb97a788)
fffffffffbc9cf80 panicsys+0x56e(fffffffffbf27cd8, fffffe004040fa18, fffffffffbc9cf90, 1)
fffffe004040fa00 vpanic+0x15c()
fffffe004040fa70 0xfffffffffb8c2361()
fffffe004040fac0 0xfffffffffbdcde25()
fffffe004040faf0 brand_solaris_setbrand+0x5a(fffffe2dbc0d0fc0, ffffffffc00d7940)
fffffe004040fb10 s10_brand`s10_setbrand+0x18(fffffe2dbc0d0fc0)
fffffe004040fb50 brand_setbrand+0x112(fffffe2dbc0d0fc0, 1)
fffffe004040fdd0 exec_common+0xc65(8047ff4, 8047fe8, 0, 2)
fffffe004040fe60 exec_init+0x28a(fffffe2da0fdf2f8, fffffe2d754f4420)
fffffe004040fe80 start_init_common+0xf5()
fffffe004040fee0 zone_start_init+0x8c()
fffffe004040fef0 thread_start+0xb()
````

--- a/usr/src/uts/common/os/brand.c
+++ b/usr/src/uts/common/os/brand.c
@@ -21,6 +21,7 @@
 /*
  * Copyright (c) 2006, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright 2019 Joyent, Inc.
+ * Copyright 2026 OmniOS Community Edition (OmniOSce) Association.
  */
 
 #include <sys/kmem.h>
@@ -641,20 +642,18 @@ brand_solaris_cmd(int cmd, uintptr_t arg1, uintptr_t arg2, uintptr_t arg3,
 void
 brand_solaris_copy_procdata(proc_t *child, proc_t *parent, struct brand *pbrand)
 {
-	brand_proc_data_t	*spd;
-
 	ASSERT(parent->p_brand == pbrand);
 	ASSERT(child->p_brand == pbrand);
 	ASSERT(parent->p_brand_data != NULL);
-	ASSERT(child->p_brand_data == NULL);
+	ASSERT(child->p_brand_data != NULL);
 
 	/*
-	 * Just duplicate all the proc data of the parent for the
-	 * child
+	 * Duplicate all the proc data of the parent for the child. The
+	 * child's brand data was already allocated by brand_setbrand()
+	 * since this brand declares a b_data_size.
 	 */
-	spd = kmem_alloc(sizeof (brand_proc_data_t), KM_SLEEP);
-	bcopy(parent->p_brand_data, spd, sizeof (brand_proc_data_t));
-	child->p_brand_data = spd;
+	bcopy(parent->p_brand_data, child->p_brand_data,
+	    sizeof (brand_proc_data_t));
 }
 
 static void
@@ -1179,22 +1178,24 @@ brand_solaris_proc_exit(struct proc *p, struct brand *pbrand)
 	ASSERT(p->p_brand == pbrand);
 	ASSERT(p->p_brand_data != NULL);
 
-	/* upon exit, free our proc brand data */
-	kmem_free(p->p_brand_data, sizeof (brand_proc_data_t));
-	p->p_brand_data = NULL;
+	/*
+	 * Nothing to do here. The proc brand data must remain in place
+	 * until the remaining LWP has been cleaned up via the b_freelwp
+	 * hook; it is then freed by brand_clearbrand() when the process
+	 * is reaped.
+	 */
 }
 
 void
 brand_solaris_setbrand(proc_t *p, struct brand *pbrand)
 {
+	ASSERT(MUTEX_HELD(&p->p_lock));
 	ASSERT(p->p_brand == pbrand);
-	ASSERT(p->p_brand_data == NULL);
 
 	/*
-	 * We should only be called from exec(), when we know the process
-	 * is single-threaded.
+	 * The proc brand data was allocated by brand_setbrand() before
+	 * this hook was called, whether branding is occurring as part of
+	 * fork() or exec().
 	 */
-	ASSERT(p->p_tlist == p->p_tlist->t_forw);
-
-	p->p_brand_data = kmem_zalloc(sizeof (brand_proc_data_t), KM_SLEEP);
+	ASSERT(p->p_brand_data != NULL);
 }

